### DISCUSSION TITLE
Typescript: how to filter operation when extend prisma and $allOperations

### DISCUSSION THREAD

**Discussion by meotimdihia** [prisma/prisma#20053] [Q&A]
Typescript: how to filter operation when extend prisma and $allOperations.
It works if I filter for `create` or `update` only. But I don't how to avoid duplicate code when extending both `create` and `update` too.
```ts
const xprisma = prisma
  .$extends({
    query: {
      md_chapters: {
        async $allOperations({ args, query, operation, model }) {
          const cur = await query(args)
          if (operation == "create" || operation == "update") {
            if (cur.status == "OK" || cur.status == "waiting") {
            ....
         }
     }
  })
```
![image](https://github.com/prisma/prisma/assets/300961/6a63532f-89ff-4c38-b131-6170b21d1ef6)

Also please add examples to the documentation. I think it is very useful.

**Comment by nurul3101** [comment]
Hello @meotimdihia 👋 

I am not able to understand the exact issue you are facing. Do you mind elaborating?

> Also please add examples to the documentation. I think it is very useful.

Did you had a chance to look at this [section](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/extension-examples) of examples?

**Reply by meotimdihia** [reply]
@nurul3101 I can't fix the Typescript error. 
I just want to run the script for `create` and `update` - I am converting middleware code to $extend code
Please check this video:


https://github.com/prisma/prisma/assets/300961/3e8e533f-d704-4e67-9f7f-5c1272351044

**Reply by janpio** [reply]
What form does `cur` have in these cases? Won't it be an array of values often?

**Reply by meotimdihia** [reply]
Hi, @janpio sorry for the late reply - I don't see the reply email.
```ts
          const cur = await query(args)
          if (operation == "create" || operation == "update") {
            if (cur.status == "OK" || cur.status == "waiting") {
```
`cur` is the current result. It is just a single record return from `create` or `update`, this code is copied from Middleware. I almost don't change it.

This code is using middleware, it has worked fine for a year:
```ts
const MdChaptersMiddleware = async (params, next) => {
  const cur = await next(params)

  if (
    params.model == "md_chapters" &&
    ["create", "update"].indexOf(params.action) != -1 &&
    (cur.status == "OK" || cur.status == "waiting")
  ) {
    if (params.args.data?.["chap"]) {
    ....
```

**Reply by millsp** [reply]
Hey @meotimdihia, middlewares don't have sticky type-checks per operation, so they won't raise a type error. 

Right now, it seems to me that the types are correct and this is a TypeScript limitation. You are saying that this query is extended on all operations. But not all operations will return a valid `status`. And because your line 316 is outside the if block, typescript will calculate different types for it. That means you need to move your line 316 into your `if` block below and it should work. Then return `return query(args)` for all other cases outside the if block.

**Reply by meotimdihia** [reply]
Hi @millsp , the problem is I want to avoid duplicated code, but I don't know how to handle it.

I edited the code in this file: https://github.com/prisma/prisma-client-extensions/blob/main/model-filters/script.ts 

If you have a solution please help me to edit the code: 

```TS
import { Prisma, PrismaClient } from "@prisma/client";
import { sub } from "date-fns";

const prisma = new PrismaClient({ log: ["query"] })
  .$extends({
    query: {
      post: {
        async $allOperations({ args, query, operation, model }) {
          if (operation != "create" && operation != "update") {
            return await query(args);
          }

          // 1
          // works fine but duplicated code
          if (operation == "create") {
            const cur = await query(args);
            await prisma.post.update({
              data: {
                title: "abcedf",
              },
              where: {
                id: cur.id,
              },
            });
          }

          // works fine but duplicated code
          if (operation == "update") {
            const cur = await query(args);
            await prisma.post.update({
              data: {
                title: "abcedf",
              },
              where: {
                id: cur.id,
              },
            });
          }

          // 2
          // doesn't work, we need a solution for this problem 

          if (operation == "create" || operation == "update") {
            const cur = await query(args);

            await prisma.post.update({
              data: {
                title: "abcedf",
              },
              where: {
                id: cur.id, // Property 'id' does not exist on type 'number | PayloadToResult<PostPayload<DefaultArgs>, // RenameAndNestPayloadKeys<PostPayload<DefaultArgs>>> | ... 4 more ... | Optional<...>'.
//  Property 'id' does not exist on type 'number'.ts(2339)
              },
            });
          }
        },
      },
    },
  })
  .$extends({
    model: {
      post: {
        unpublished: () => ({ published: false }),
        published: () => ({ published: true }),
        byAuthor: (authorId: string) => ({ authorId }),
        byAuthorDomain: (domain: string) => ({
          author: { email: { endsWith: `@${domain}` } },
        }),
        hasComments: () => ({ comments: { some: {} } }),
        hasRecentComments: (date: Date) => ({
          comments: { some: { createdAt: { gte: date } } },
        }),
        titleContains: (search: string) => ({ title: { contains: search } }),
      } satisfies Record<string, (...args: any) => Prisma.PostWhereInput>,
    },
  });

async function main() {
  const posts = await prisma.post.findMany({
    where: {
      AND: [
        prisma.post.published(),
        prisma.post.byAuthorDomain("prisma.io"),
        prisma.post.hasRecentComments(sub(new Date(), { weeks: 1 })),
        prisma.post.titleContains("GraphQL"),
      ],
    },
  });
}

main()
  .then(async () => {
    await prisma.$disconnect();
  })
  .catch(async (e) => {
    console.error(e);
    await prisma.$disconnect();
    process.exit(1);
  });

```

**Reply by millsp** [reply]
Right, thanks. I think I see what the problem is now. I believe this could be related to [this](https://github.com/prisma/prisma/pull/20285) issue. I'll get back to you when it is merged.

**Reply by millsp** [reply]
Hey @meotimdihia, I just gave your issue a try with some recent fixes we merged. Your's is something else, and is an improvement that we could do. I took the time to understand what this change would mean and it sounds like a good idea, so that you can have better narrowing. In some cases we could also consider this a bug, because it is surprising that it works this way. This boils down to the way TypeScript works and the concept of variance, and ideally I'd like that our users needn't worry about this.

**Reply by millsp** [reply]
That said, I am not sure how this can be fixed without impacting autocompletion for the input. Could you please create a new issue, and we'll try to schedule work for it?

**Reply by meotimdihia** [reply]
@millsp thanks, I created the issue: https://github.com/prisma/prisma/issues/20395

**Comment by meotimdihia** [comment]
Does anyone have an answer to this question?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
